vsphere/virtual_machine_config_structure.go

```diff
@@ -508,6 +509,117 @@ func flattenExtraConfig(d *schema.ResourceData, opts []types.BaseOptionValue) er
 	return d.Set("extra_config", ec)
 }
 
+// expandVAppConfig reads in all the vapp key/value pairs and returns
+// the appropriate VmConfigSpec.
+//
+// We track changes to keys to determine if any have been removed from
+// configuration - if they have, we add them with an empty value to ensure
+// they are removed from vAppConfig on the update.
+func expandVAppConfig(d *schema.ResourceData) types.BaseVmConfigSpec {
+	if !d.HasChange("vapp") {
+		return nil
+	}
+
+	var (
+		props  []types.VAppPropertySpec
+		allIds sort.StringSlice
+	)
+
+	old, new := d.GetChange("vapp")
+
+	getVAppProps := func(val interface{}) map[string]interface{} {
+		vApps := val.([]interface{})
+		if vApps != nil && len(vApps) > 0 && vApps[0] != nil {
+			vApp := vApps[0].(map[string]interface{})
+			if props, ok := vApp["properties"].(map[string]interface{}); ok {
+				return props
+			}
+		}
+		return make(map[string]interface{})
+	}
+
+	oldMap := getVAppProps(old)
+	newMap := getVAppProps(new)
+
+	for k := range oldMap {
+		allIds = append(allIds, k)
+	}
+
+	for k := range newMap {
+		if _, ok := oldMap[k]; !ok {
+			allIds = append(allIds, k)
+		}
+	}
+	sort.Sort(allIds)
+
+	for i, k := range allIds {
+		_, isOld := oldMap[k]
+		_, isNew := newMap[k]
+
+		if isNew {
+			prop := types.VAppPropertySpec{
+				ArrayUpdateSpec: types.ArrayUpdateSpec{
+					Operation: types.ArrayUpdateOperationEdit,
+				},
+				Info: &types.VAppPropertyInfo{
+					Key:   int32(i) + 1,
+					Id:    k,
+					Value: newMap[k].(string),
+				},
+			}
+			props = append(props, prop)
+		} else if isOld {
+			prop := types.VAppPropertySpec{
+				ArrayUpdateSpec: types.ArrayUpdateSpec{
+					Operation: types.ArrayUpdateOperationEdit,
+				},
+				Info: &types.VAppPropertyInfo{
+					Key:   int32(i) + 1,
+					Id:    k,
+					Value: "",
```

Comment:
Just noticed this here - the original comment said that there was problems zeroing out a property - is this code non-functional, or were you referring to something else (maybe deleting a key completely)?

Reply:
The issues with zeroing out properties are due to the fact that at least in version 5.5, the vSphere API doesn't actually do anything when it receives an empty string for a vApp property value. We've gotten around this by changing the empty string to " ", and the corresponding portion of the new code should work properly now.

Reply:
Awesome! This is great info actually as I can now update some of my code on the work I was doing on govc. Good find! :+1: